packages/dev/s2-docs/src/NavigationSuspense.tsx

```diff
@@ -0,0 +1,121 @@
+'use client';
+
+import type {Page} from '@parcel/rsc';
+import {PageSkeleton} from './PageSkeleton';
+import React, {Suspense, use, useSyncExternalStore} from 'react';
+
+let navigationPromise: Promise<void> | null = null;
+let navigationResolve: (() => void) | null = null;
+let targetPathname: string | null = null;
+let listeners = new Set<() => void>();
+let cachedSnapshot: {promise: Promise<void> | null, pathname: string | null} = {promise: null, pathname: null};
+
+function subscribe(callback: () => void) {
+  listeners.add(callback);
+  return () => listeners.delete(callback);
+}
+
+function getSnapshot() {
+  if (cachedSnapshot.promise !== navigationPromise || cachedSnapshot.pathname !== targetPathname) {
+    cachedSnapshot = {promise: navigationPromise, pathname: targetPathname};
+  }
+  return cachedSnapshot;
+}
+
+export function setNavigationLoading(loading: boolean, pathname?: string) {
+  if (loading) {
+    targetPathname = pathname || null;
+    navigationPromise = new Promise(resolve => {
+      navigationResolve = resolve;
+    });
+    listeners.forEach(callback => callback());
+  } else {
+    if (navigationResolve) {
+      navigationResolve();
+      navigationResolve = null;
+      navigationPromise = null;
+      listeners.forEach(callback => callback());
+    }
+  }
+}
+
+function normalizePathname(urlOrPathname: string, publicUrlPrefix: string): string {
+  let pathname: string;
+  if (urlOrPathname.startsWith('http://') || urlOrPathname.startsWith('https://')) {
+    try {
+      pathname = new URL(urlOrPathname).pathname;
+    } catch {
+      pathname = urlOrPathname;
+    }
+  } else {
+    pathname = urlOrPathname;
+  }
+  
+  const [basePathname] = pathname.split('?');
+  const [cleanPathname] = basePathname.split('#');
```

Comment:
If you parsed it with `new URL` then this shouldn't happen. If it didn't start with `http` then you could assume it's a relative url to the current location? `new URL(urlOrPathname, location.href)`